### PR TITLE
feat: (opt-in): terminate handling of work when the request has already timed out

### PR DESCRIPTION
**Pull request by jrmfg**
Overhead-free (or at least very cheap).

The “timeout” gunicorn config means drastically different things for
sync and non-sync workers:

Workers silent for more than this many seconds are killed and restarted.

Value is a positive number or 0. Setting it to 0 has the effect of
infinite timeouts by disabling timeouts for all workers entirely.

Generally, the default of thirty seconds should suffice. Only set this
noticeably higher if you’re sure of the repercussions for sync workers.
For the non sync workers it just means that the worker process is still
communicating and is not tied to the length of time required to handle a
single request.

So.  For cases where threads = 1 (user set or our defaults), we’ll use
the sync worker and let the regular timeout functionality do its thing.

For cases where threads > 1, we’re using the gthread worker, and timeout
means something completely different and not really user-observable.  So
we’ll leave the communication timeout (default gunicorn “timeout”) at 30
seconds, but create our own gthread-derived worker class to use instead,
which terminates request handling (with no mind to gunicorn’s “graceful
shutdown” config), to emulate GCF 1st gen.

The arbiter spawns these workers, so we have to maintain some sort of
global timeout state for us to read in our custom gthread worker.

In the future, we should consider letting the user adjust the graceful
shutdown seconds.  But the default of 30 seems like it’s worked fine
historically, so it’s hard to argue for changing it.  IIUC, this means
that on gen 2, there’s a small behavior difference for the sync workers
compared to gen 1, in that gen 2 sync worker workloads will get an extra
30 seconds of timeout to gracefully shut down.  I don’t think monkeying
with this config and opting-in to sync workers is very common, though,
so let’s not worry about it here; everyone should be on the gthread path
outlined above.

### CODE REVIEW

File: src/functions_framework/_http/gunicorn.py [hunk 1]
```diff
@@ -38,3 +64,13 @@ def load_config(self):
...
+
+
+class GThreadWorkerWithTimeoutSupport(ThreadWorker):  # pragma: no cover
+    def handle_request(self, req, conn):
+        with ThreadingTimeout(TIMEOUT_SECONDS):
+            super(GThreadWorkerWithTimeoutSupport, self).handle_request(req, conn)
+
+
+def _is_truthy(s):  # pragma: no cover
+    return str(s).lower() in ("yes", "y", "1", "yeah", "true", "t")
```
Comment (HKWinterhalter) [inline src/functions_framework/_http/gunicorn.py:76]: nit: is this necessary? It seems like we'd be fine to just accept "true" if the env var / feature is documented

File: tests/test_functions/timeout/main.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import logging
+import time
+
+logger = logging.getLogger(__name__)
+
+
+def function(request):
+    # sleep for 1200 total ms (1.8 sec)
```
Comment (HKWinterhalter) [inline tests/test_functions/timeout/main.py:8]: nit: is 1.8s a typo?

File: tests/test_timeouts.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            if time.perf_counter() - start_time >= timeout:
+                raise TimeoutError(
+                    "Waited too long for port {} on host {} to start accepting "
+                    "connections.".format(port, host)
+                ) from ex
+
+
+@pytest.mark.skip
+def _wait_for_no_listen(host, port, timeout=10):
+    # Used in tests to make sure that the
```
Comment (HKWinterhalter) [inline tests/test_timeouts.py:186]: nit: unfinished phrase

File: tests/test_timeouts.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    assert result.status_code == 200
+
+
+@pytest.mark.skipif("platform.system() == 'Windows'")
+@pytest.mark.skipif("platform.system() == 'Darwin'")
+@pytest.mark.slow_integration_test
+def test_timeout_but_not_threaded_timeout_enabled_does_not_kill(monkeypatch):
+    monkeypatch.setenv("CLOUD_RUN_TIMEOUT_SECONDS", "1")
+    monkeypatch.setenv("THREADED_TIMEOUT_ENABLED", "false")
```
Comment (HKWinterhalter) [inline tests/test_timeouts.py:75]: should there be a case where THREADED = false, but (sync) WORKER still fails due to timeout?

i.e.: else:
            self.options["timeout"] = TIMEOUT_SECONDS

File: src/functions_framework/_http/gunicorn.py (hunk 1 above)
Comment (jrmfg) [reply src/functions_framework/_http/gunicorn.py:76]: sure, let's just yank it

**Review by HKWinterhalter — APPROVED** [review]

**Comment by HKWinterhalter** [comment]
Looks great!

**Pull request merged**